### ISSUE TITLE
no method 'setSpout' after upgrade to redstorm 0.4.0 and storm 0.6.2

### ISSUE THREAD

**Issue by stonegao** [colinsurprenant/redstorm#15]
I upgraded to redstorm 0.4.0 today for our project (previously its based on redstorm 0.2.1). When running the topology locally the following exception occurs:

<pre>
0    [main] ERROR org.apache.zookeeper.server.NIOServerCnxn  - Thread Thread[main,5,main] died
org.jruby.exceptions.RaiseException: (NameError) no method 'setSpout' for arguments (org.jruby.RubyString,redstorm.storm.jruby.JRubySpout,org.jruby.RubyFixnum) on Java::BacktypeStormTopology::TopologyBuilder
    at EventTopology.start(/Users/stone/.rvm/gems/jruby-1.6.6/gems/redstorm-0.4.0/lib/red_storm/simple_topology.rb:110)
    at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1612)
    at RedStorm::SimpleTopology.start(/Users/stone/.rvm/gems/jruby-1.6.6/gems/redstorm-0.4.0/lib/red_storm/simple_topology.rb:108)
    at #<Class:0x77b4ee5e>.main(/Users/stone/.rvm/gems/jruby-1.6.6/gems/redstorm-0.4.0/lib/red_storm/topology_launcher.rb:43)
</pre>

**Comment by colinsurprenant** [comment]
Hi Stone,

After upgrading, did you clean your target/ directory and reinstall with "redstorm install" ?
If that does not help, would it be possible to submit a topology which recreates this problem? 

Colin

**Comment by stonegao** [comment]
Thanks for your suggestion. I only did `redstorm install` and didn't clean the target/ directory.  After clean & reinstall the exception disappeared but a new exception occurs: 

<pre>
1    [main] ERROR org.apache.zookeeper.server.NIOServerCnxn  - Thread Thread[main,5,main] died
org.jruby.exceptions.RaiseException: (LoadError) load error: bundler/definition -- java.lang.StackOverflowError: null
    at #<Class:0x2e4e76b4>.definition(/Users/stone/.rvm/gems/jruby-1.6.6/gems/bundler-1.0.22/lib/bundler.rb:138)
    at #<Class:0x2e4e76b4>.load(/Users/stone/.rvm/gems/jruby-1.6.6/gems/bundler-1.0.22/lib/bundler.rb:126)
    at #<Class:0x2e4e76b4>.setup(/Users/stone/.rvm/gems/jruby-1.6.6/gems/bundler-1.0.22/lib/bundler.rb:110)
</pre> 


I googled, seems that it's a JRuby bug now. 

http://jira.codehaus.org/browse/JRUBY-6400
http://jira.codehaus.org/browse/JRUBY-6407

**Comment by colinsurprenant** [comment]
I think you are right, it looks like its the same problem. Have you tried the suggested fix per http://jira.codehaus.org/browse/JRUBY-6407?focusedCommentId=290552&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-290552

If you use rvm, this file is in ~/.rvm/rubies/jruby-1.6.6/lib/ruby/site_ruby/1.8/rubygems/defaults/jruby.rb and you can simply move the line with require 'jruby/util' to the top of the file.

**Comment by stonegao** [comment]
Thanks for your advice.

tried the fix but not work. I also tried install jruby-1.6.6-head not work too.

**Comment by colinsurprenant** [comment]
Can you gist a topology which reproduces this problem?

**Comment by stonegao** [comment]
I tried change the pom.xml in redstorm and build a redstorm gem that depends on jruby-1.6.5.1 and it works as before now. 

The topology is just like the examples in redstorm, except that I added one line `require boot` and `boot.rb` was mentioned in https://github.com/colinsurprenant/redstorm/issues/7 : 

``` ruby
require 'rubygems'

# Set up gems listed in the Gemfile.
PROJECT_ROOT = File.expand_path('../../..', __FILE__)

gemfile = File.join(PROJECT_ROOT, 'Gemfile')
begin
  ENV['BUNDLE_GEMFILE'] = gemfile
  require 'bundler'
  Bundler.setup
rescue Bundler::GemNotFound => e
  STDERR.puts e.message
  STDERR.puts "Try running `bundle install`."
  exit!
end if File.exist?(gemfile)

Bundler.require(:default)
```

**Comment by colinsurprenant** [comment]
Ok, so this confirms that this problem is in fact related to the JRuby 1.6.6 + rubygems bug http://jira.codehaus.org/browse/JRUBY-6407 - the fix will come with the next JRuby release.

Otherwise, requiring bundler from your topology will only work in local mode. When run remotely on a cluster, the gems will need to be installed into target/gems/ and part of the generated target/cluster-topology.jar

I haven't tried it with bundler yet, but I have tested cluster + gems using standard rubygems by installing the gems into target/gems/ with the command gem install <the gem> --install-dir target/gems/ --no-ri --no-rdoc

I opened ticket #16 for a better bundler integration.
